### DISCUSSION TITLE
Switch from Rosetta2 to native ARM install

### DISCUSSION THREAD

**Discussion by josherrickson** [Homebrew/discussions#545] [Getting started] [answered]
I have an M1 Mac that I installed brew through Rosetta2. I'd like to try switching to the ARM native version. Is there a best practices approach? Do I need to completely remove everything I installed with it and reinstall? Can I have both versions installed and gradually transition?

**Comment by richiksc** [comment, marked as answer]
Yes, assuming your Rosetta 2 installation was installed to `/usr/local` as recommended, if you run the install script now, it'll install the ARM native installation to `/opt/homebrew`. This was part of the reason why two different prefixes were chosen.

Then, make sure `/opt/homebrew/bin` is added to your `PATH` prepended before `/usr/local/bin`. For example, in your `.zprofile` or `.zshrc`:

```
eval $(/opt/homebrew/bin/brew shellenv)
```

Then you can alias `/usr/local/bin/brew` to `ibrew` or `x64brew` or something similar:
```
alias ibrew='arch -x86_64 /usr/local/bin/brew`
```

Running brew from your terminal should run the ARM version of brew by default. You can verify by using `brew config`.
Running `ibrew` should run the Rosetta version.

Running a formula will prefer the ARM native version since `/opt/homebrew/bin` comes first in your path. To run the Rosetta version of a formula that you've also installed for ARM, you'd need to prefix it with `/usr/local/bin`. If you don't have both versions, your shell should select which one automatically. When you've completely transitioned over, if you run `uninstall.sh` under Rosetta, it should uninstall the Intel version of Homebrew only.

```
arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall.sh)"
```

**Reply by josherrickson** [reply]
Ran into a bit of trouble since I had fish (installed through x86 brew) as my default shell. (`uname -m` was returning x86_64.) Solution was to change the default shell to `/bin/bash` before installing. After that, everything worked great, thanks!

**Reply by richiksc** [reply]
Yes, you need to install ARM homebrew in an ARM shell. I highly recommend you use an ARM shell as your default shell, and just drop into an x86_64 shell if you need to do extended work with Rosetta, or just single prefix commands with `arch -x86_64`. `fish` is available as an ARM native build on ARM Homebrew as well.

And if you want to drop into a Rosetta shell, it's simple as running`arch -x86_64 /usr/local/bin/fish`. You could set up an alias called `x64shell` or `rosetta`.

Most formulae are available for ARM now so you shouldn't need to use Rosetta much, the transition should be fairly easy. I've never even installed Rosetta Homebrew.

**Comment by brokosz** [comment]
What's the best way to cleanup the installed packages? My thought was to `brew bundle dump` then uninstall everything?, re-run the brew install natively, and reapply the bundle. Not sure if there's a recommended set of steps here or better way to do this.

**Reply by carlocab** [reply]
That sounds pretty much like how I would do it.

**Reply by sumater** [reply]
But what if I forget to `bundle dump` before uninstalling? that's what I did.

**Reply by gromgit** [reply]
Answered here: https://github.com/orgs/Homebrew/discussions/417#discussioncomment-10856194

**Comment by brokosz** [comment]
Okay - that worked great. Here are the steps I followed. 

## Migrating to native (ARM) Homebrew

These steps will remove all formulas and taps, preserving installed casks and Mac App Store apps. If you installed a new shell via Homebrew, you will have to change your shell in Terminal/iTerm before step 4 to a natively installed shell. 

1. Run `brew bundle dump`
1. Edit `brewfile` to split it into two new files: remove casks and MAS apps (if mas installed), leaving all installed formulas and related taps in a `reinstall` file. Then save the casks and MAS apps in a `cleanup` file. 
1. Run `brew bundle cleanup -f --file <path to edited CLEANUP brewfile>`
1. Run Homebrew uninstaller `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"`
1. Reinstall homebrew using the usual command `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
1. Update your `PATH` to add the new Homebrew path entries
1. Run `brew bundle --file <path to edited REINSTALL brewfile>` to reinstall formulas and taps
1. Update profile, scripts, etc. as needed

**Reply by brokosz** [reply]
... and I missed a step in here for repointing the installed casks so now I'm off reinstalling all of those ☹️

**Reply by ilovecomputers** [reply]
Does anyone have details of how to repoint the installed casks?

**Reply by SMillerDev** [reply]
Reinstalling is the best way because otherwise you're stuck with Intel versions of them. In general you can use `brew install --cask --adopt` to adopt installed casks into homebrew though.
